rfcs/rfc-0003-discriminators.md

```diff
@@ -0,0 +1,262 @@
+- Feature Name: Discriminators
+- Start Date: 06/05/2023
+- RFC PR: https://github.com/revoltchat/rfcs/pull/3
+- Tracking Issue: https://github.com/revoltchat/backend/issues/247
+- Status: draft
+
+# Summary
+
+Revolt should introduce discriminators which serve as part of a semi-hidden extension to
+usernames which allows multiple users to have the same username, the discriminator would
+not be visible in most circumstances except when viewing a person's profile and if you
+were to add said person by their username.
+
+# Motivation
+
+This RFC aims to move forwards [prior discussion on the feature suggestions forum](https://github.com/orgs/revoltchat/discussions/61),
+it stands to solve multiple problems when implemented correctly:
+
+- Users should in most, if not all, cases be able to pick their chosen username.
+- This should curb any incentive to sell / steal accounts for "rare usernames".
+
+  Certain "rare" discriminators can also be prohibited.
+
+- Allow users who do not want to be easily discoverable to stay hidden.
+
+  Just a 4-digit discriminator provides 10,000 unique users.
+  Expanding on this, we could for example, let users hide discriminators in servers and similar privacy features.
+
+# Guide-level explanation
+
+Revolt would be switching to a new username system made up of three parts:
+
+```
+myusername#1234
+^ your chosen name (restricted character set)
+          ^ separator between chosen name and discriminator
+           ^ discriminator (4-digits in range 0001 to 9999)
+```
+
+Users will also be able to set an optional display name which overrides their username
+in various parts of the user interface.
+
+Usernames use a restricted character set of any Unicode alphabet excluding some lookalike
+characters from the Cyrillic alphabet, if you username does not currently fit the specification,
+then it will be automatically santised upon upgrade and copied to your display name.
+
+Discriminators for new users are generated randomly on sign up while existing users will
+get a randomly generated discriminators upon the server software being upgraded.
+
+> Revolt treats usernames as case-insensitive so a full username such as "Jason#1234" is
+> effectively the same as "jasON#1234" but only "Jason#1234" will display in the UI.
+
+Users may:
+
+- Change the case of their username at any moment, such as from "Jason" to "jasON".
+
+  This will never affect their discriminator.
+
+- Change their username, but may have their discriminator regenerated if there is a conflict.
+
+  Such as if you are changing from "Jason#1234" to "Phil", but "Phil#1234" is taken.
```

Comment:
Just pulling this out of the document here.
This RFC is pretty much ready to go now I believe.

Just to summarise, the current proposal is:
- 4 digit discriminator (0001 to 9999)
- Discriminators will be randomly assigned for all users
- Restrict usernames further to (Unicode alphabet + digits + "_-.") excluding some lookalike characters (this will affect 10% of usernames on platform).
- Display names to be added
- Uncompliant usernames will be sanitised and originals moved to display name
- Discriminator will be regenerated on conflict

Implementation will likely have much stricter rate limits for changing username.

Reply:
are there any plans to restrict "valuable" discriminators? e.g 0001, 0420, 6969, etc

Reply:
The following list has been decided:
````
0001
1111
2222
3333
4444
5555
6666
7777
8888
9999
```